opm/simulators/wells/GasLiftSingleWell.hpp

```diff
@@ -0,0 +1,235 @@
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef OPM_GASLIFT_SINGLE_WELL_HEADER_INCLUDED
+#define OPM_GASLIFT_SINGLE_WELL_HEADER_INCLUDED
+
+#include <opm/models/utils/propertysystem.hh>
+#include <opm/models/utils/parametersystem.hh>
+#include <opm/core/props/BlackoilPhases.hpp>
+#include <opm/output/data/Wells.hpp>
+#include <opm/parser/eclipse/EclipseState/Schedule/Schedule.hpp>
+#include <opm/parser/eclipse/EclipseState/Schedule/Well/Well.hpp>
+#include <opm/simulators/utils/DeferredLogger.hpp>
+#include <opm/parser/eclipse/EclipseState/Schedule/GasLiftOpt.hpp>
+// NOTE: StandardWell.hpp includes ourself (GasLiftSingleWell.hpp), so we need
+//   to forward declare StandardWell for it to be defined in this file.
+namespace Opm {
+    template<typename TypeTag> class StandardWell;
+}
+#include <opm/simulators/wells/StandardWell.hpp>
+
+#include <opm/simulators/wells/WellStateFullyImplicitBlackoil.hpp>
+#include <opm/core/props/BlackoilPhases.hpp>
+#include <opm/simulators/wells/GasLiftWellState.hpp>
+
+#include <algorithm>
+#include <cassert>
+#include <cmath>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <optional>
+#include <string>
+#include <tuple>
+#include <vector>
+#include <utility>
+#include <fmt/format.h>
+
+namespace Opm
+{
+    template<class TypeTag>
+    class GasLiftSingleWell
+    {
+        using Simulator = GetPropType<TypeTag, Properties::Simulator>;
+        using WellState = WellStateFullyImplicitBlackoil;
+        using StdWell = Opm::StandardWell<TypeTag>;
+        using GLiftWellState = Opm::GasLiftWellState<TypeTag>;
+        // TODO: same definition with WellInterface, and
+        //  WellStateFullyImplicitBlackoil eventually they should go
+        //  to a common header file.
+        static const int Water = BlackoilPhases::Aqua;
+        static const int Oil = BlackoilPhases::Liquid;
+        static const int Gas = BlackoilPhases::Vapour;
+        static constexpr double ALQ_EPSILON = 1e-8;
+        struct OptimizeState;
+        class Stage2State;
+    public:
+        GasLiftSingleWell(
+            const StdWell &std_well,
+            const Simulator &ebos_simulator,
+            const SummaryState &summary_state,
+            DeferredLogger &deferred_logger,
+            WellState &well_state
+        );
+        struct GradInfo;
+        std::optional<GradInfo> calcIncOrDecGradient(
+            double oil_rate, double gas_rate, double alq, bool increase) const;
+        std::pair<double, double> getStage2Rates();
+        const WellInterface<TypeTag> &getStdWell() const { return std_well_; }
+        bool hasStage2Rates();
+        std::unique_ptr<GLiftWellState> runOptimize();
```

Comment:
My experience from a period of heavy `std::shared_ptr<T>` abuse is that the public api should only expose `GliftWellState`, `const GliftWellState&` and possibly `const GLiftWellState *` - by using `std::xxx_ptr<T>` in the API you impose restrictions on downstream which is really not necessary.